### ISSUE TITLE
Theme CSS loading in wrong order due to 'every_page_weight' and backdrop_css_alter()

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#4942]
**Description of the bug**

This is in response [to this issue](https://github.com/backdrop-contrib/bootstrap_lite/issues/13). The problem results in theme CSS and JS being loaded out of order.

The issue revolves around the key `'every_page_weight'`, which is added by `backdrop_add_css()` and `backdrop_add_js()`, and is used by `backdrop_get_css()` and `backdrop_get_js()` to determine the proper ordering of loaded CSS functions.

The problem arises when developers use `hook_css_alter()` (or the JS equivalent) to conditionally add CSS, as, for example, is done in the Bootstrap Lite theme (see issue linked above). If `$options['every_page_weight']` is not set properly in the `hook_css_alter()` implementation, then the conditionally added CSS gets loaded later than it's supposed to. In the issue above, the net result is that the conditional CSS of the base theme overrides any ostensible CSS overrides in subthemes.

**Steps To Reproduce**

To reproduce the behavior:
1. Enable a theme with conditional loading of CSS, like Bootstrap Lite.
2. Set it to load Bootstrap from the CDN.
3. In browser developer tools, look at the order in which CSS and JS files are loaded.

**Actual behavior**

The conditional files added by the theme will come late in the load order, thereby overriding subthemes and/or modules.

**Expected behavior**

Subthemes can override the base theme in the usual way.

**Possible solution**

One solution is fairly simple: themers should set the value of this key appropriately  in `hook_css_alter()`. So that they know this, document what it should be set to, based on the value of `$options['every_page']` (it's currently undocumented).

But perhaps a cleaner solution would be to remove the setting of `$options['every_page_weight']` from `backdrop_add_css()`, and instead set its value in `backdrop_get_css()`, after the call to `backdrop_alter()`, but before the call to `backdrop_sort()`, which appears to be the only place that it's used. Then themers don't have to worry about it at all, and that key becomes completely localized in scope.

**Additional information**

Add any other information that could help, such as:
- Backdrop CMS version: 1.18.1
-

**Comment by indigoxela** [comment]
@bugfolder many thanks for digging into this.

> themers should set the value of this key appropriately in hook_css_alter(). So that they know this, document what it should be set to, based on the value of $options['every_page'] (it's currently undocumented).

I agree, we should be able to improve documentation re "every_page_weight".

> perhaps a cleaner solution would be to remove the setting of $options['every_page_weight'] from backdrop_add_css(), and instead set its value in backdrop_get_css(), after the call to backdrop_alter(), but before the call to backdrop_sort(), ...

I suspect, such a change has the potential to break existing stuff.

**Comment by bugfolder** [comment]
It has the potential to break existing stuff, but it's worth looking into the specific circumstances where it might break existing stuff. I submit that some stuff, perhaps, _should_ be broken.

So first, let's look at core. Doing a search for `every_page_weight` in core reveal that it only shows up 4 times, the places described above, all in `common.inc`:

* in `backdrop_add_css()`, where it is added to the `$options` array;
* in `backdrop_add_js()`, ditto;
* in `backdrop_get_css()`, where it is used in the sorting;
* in `backdrop_get_js()`, ditto.

Currently, any implementations of `hook_css_alter()` that set `'every_page' => TRUE` but do _not_ set the undocumented `'every_page_weight'` key will load CSS out of the documented order. The current order, [documented for `backdrop_add_css`](), is (in the section on the `'weight'` key):

> * First by group.
> * Then by the `'every_page'` flag, with TRUE coming before FALSE.
> * Then by weight.
> * Then by the order in which the CSS was added. For example, all else being the same, a stylesheet added by a call to `backdrop_add_css()` that happened later in the page request gets added to the page after one for which `backdrop_add_css()` happened earlier in the page request.

But this is not actually the current behavior! Here's the current behavior:

> * First by group.
> * Then by the `'every_page'` flag, with TRUE coming before FALSE, **unless a theme or module you are using added CSS or JS and set `'every_page' => TRUE` but did not set the undocumented `'every_page_weight'` flag appropriately, in which case in their module, `'every_page'` is ignored with respect to ordering, though still respected with respect to loading, and their CSS might load after yours. The only way to know if a theme or module you are using did this is to either inspect the loading order in your browser developer tools or examine the code of all modules or themes that you rely upon. **
> * Then by weight.
> * Then by the order in which the CSS was added. For example, all else being the same, a stylesheet added by a call to `backdrop_add_css()` that happened later in the page request gets added to the page after one for which `backdrop_add_css()` happened earlier in the page request.

So, yeah, if a module or theme relied on the exceptional behavior in bold above, changing things per my second suggestion would break that behavior. But I respectfully suggest that such behavior **should** be broken, because I think it is more often going to be confusing and hard to diagnose (witness the long thread in the Bootstrap Lite issue).

And even if we say "no, we have to keep the current behavior," we certainly need to document the _actual_ behavior of `backdrop_add_css()`, which would be something along the lines of my bold text above.

I looked through all core implementations of `hook_css_alter()` and `hook_js_alter()`. There are no implementations that set the `'every_page'` flag, so I don't think this change would alter any behavior in core.

It would only potentially change the behavior of any existing themes or modules that implement `hook_css_alter()` or `hook_js_alter()`, have `'every_page' => TRUE`, and didn't set `'every_page_weight'`. 

That's probably a small group.

Within that group, anyone affected by the change would fall into two categories:

1. People who can't understand why the load order differs from what's documented; would be happy because load order now matches current documentation.

2. People who relied on the weird load order: "I don't understand why it's doing this, but I'll do some weird patch that overrides the weird load order." Their patch might no longer work.

Making the change might break the patches of group 2, but those are the cases for which I think breaking stuff might be warranted.

And I would consider that preferable to revising the documentation so as to enshrine the current behavior and add one more key that developers need to be aware of.

Last (and I apologize for the length), I _suspect_ that `'every_page_weight'` was never intended to be exposed to sub-theme or module developers; its intention was to be a private variable, used only for sorting. Keeping it private is, I think, an overall cleaner solution.

**Comment by indigoxela** [comment]
>  if a module or theme relied on the exceptional behavior in bold above, changing things per my second suggestion would break that behavior. But I respectfully suggest that such behavior should be broken

I belief, the preferred behavior is *not* to break contrib. But I'm not the one to decide that, I just gather that from the Backdrop [philosophy statement](https://backdropcms.org/philosophy). Maybe one of the core committers can clarify, how to proceed here.

**Comment by quicksketch** [comment]
If I understand this correctly, would using this logic solve the problem?

* When processing the css_alter and js_alter hooks...
* If "every_page" is set to `TRUE` AND "weight" is set AND "every_page_weight" is NOT set 
* Set "every_page_weight" to the value of the "weight"

And then everything else works and sorts as it does now. "every_page_weight" just automatically copies "weight" if one is set but not the other, and we continue treating it as an "internal" key.

**Comment by bugfolder** [comment]
@quicksketch, I don't think that's quite it. `every_page_weight` only takes on the values {+1,-1}, whereas `weight` can be anything. But I think the logic of "if `every_page_weight` isn't set, set it to what it should have been set to in order to get the documented behavior" would work, and would give the most reasonable behavior.

I think that would be (in `backdrop_get_css()`, just before the sort):

```
foreach ($css as $key => &$item) {
  $item['options']['every_page_weight'] = $item['options']['every_page_weight'] ?? $item['options']['every_page'] ? -1 : 1;
}```

And ditto for `backdrop_get_js()`.

If there's agreement, I'd be happy to prepare a PR.

**Comment by indigoxela** [comment]
> If there's agreement, I'd be happy to prepare a PR.

@bugfolder your analysis absolutely makes sense to me. Just go for it! :smiley:

**Comment by bugfolder** [comment]
[PR ready for review.](https://github.com/backdrop/backdrop/pull/3532) I've tested it with Bootstrap Lite and a subtheme, and it gives the proper CSS loading order.

**Comment by indigoxela** [comment]
I tried to test this like that:

In a custom theme I added in the template.php file:
```
function MYTHEME_css_alter(&$css) {
  $filepath = backdrop_get_path('theme', 'MYTHEME') . '/foocustom.css';
  $css[$filepath] = array(
    'data' => $filepath,
    'type' => 'file',
    'every_page' => TRUE,
    'media' => 'all',
    'preprocess' => TRUE,
    'group' => CSS_THEME,
    'browsers' => array(),
  );
}
```
Content of foocustom.css:
```
body {
  color: green;
}
```
And then I inspected the order of css files appearing in source, with and without patch applied.

Without it foocustom.css always came last, no matter to what value "every_page" was set. With the change the css file got the expected position in the list of css files - according to how it is documented (...Then by the 'every_page' flag, with TRUE coming before FALSE)

However, I realized that switching the change on/off made the green color for body take effect or do not take effect, so this has (as expected) impact on existing implementations of hook_css_alter().

FTR: I did add conditional css files in custom themes, but never did it with that hook, but usually with template_preprocess_page(). So I can't estimate, if the change will affect many or no sites... leaning towards "probably none to very few". But maybe I'm wrong.

I still support that change, because if fixes odd behavior that contradicts our official documentation, so "works for me".
